### ISSUE TITLE
Provide safe allocation API for varsized structs

### ISSUE THREAD

**Issue by alex** [rust-lang-nursery/rust-bindgen#1423]
Currently if you have a varsized struct (`struct { size_t length; uint64_t data[] };`), you're responsible for figuring out how to allocate it yourself, which can be messy and `unsafe`.

Would it be reasonable for the generated Rust type for that struct to have an API for allocating it, with a certain length (NB, the user would still be responsible for filling in the `length` field)?

This would be beneficial by a) being more ergonomic, and more importantly b) consistently handling things like integer overflows to make code safer!

**Comment by emilio** [comment]
I agree this would be useful, though we'd need heuristics to know where to provide it, since you can also use zero-sized arrays to separate bitfields for example, or stuff like that, IIRC.

Probably a good heuristic would be something like "only one zero-size array, at the end of the struct".

I'm happy to mentor this or what not :)

**Comment by emilio** [comment]
So, the main place to look at would be `impl CodeGenerator for CompInfo`. That's the code generation for both structs and unions, and there you should have everything you need.

In particular, we already generate a few `methods` for C++, so you should add stuff there. Probably adding a new option you can opt-in would be preferred.

You probably need a condition like "no funky stuff, not an union, and all my fields are pod, and last field is a var-sized array of something we understand.". Maybe it may make sense to opt into it per-type.

**Comment by emilio** [comment]
That's all in `src/codegen/mod.rs`.
